src/net.cpp

```diff
@@ -725,12 +723,29 @@ int CNetMessage::readHeader(const char *pch, unsigned int nBytes)
         hdrbuf >> hdr;
     }
     catch (const std::exception&) {
+        LogPrint("net", "Header error: failed to deserialize\n");
         return -1;
     }
 
-    // reject messages larger than MAX_SERIALIZED_COMPACT_SIZE
-    if (hdr.nMessageSize > MAX_SERIALIZED_COMPACT_SIZE)
-            return -1;
+    // reject messages with invalid start
+    if (memcmp(hdr.pchMessageStart, Params().MessageStart(), MESSAGE_START_SIZE) != 0) {
+        LogPrint("net", "Header error: invalid start (%s)\n", SanitizeString(std::string(hdr.pchMessageStart, hdr.pchMessageStart + MESSAGE_START_SIZE)));
+        return -1;
+    }
+
+    // reject messages with invalid command
+    if (!hdr.HasValidCommand())
+    {        
+        LogPrint("net", "Header error: invalid command (%s)\n", SanitizeString(hdr.GetCommand()));
+        return -1;
+    }
+
+    // reject messages larger than MAX_PROTOCOL_MESSAGE_LENGTH or MAX_SERIALIZED_COMPACT_SIZE
+    if (hdr.nMessageSize > MAX_PROTOCOL_MESSAGE_LENGTH || hdr.nMessageSize > MAX_SERIALIZED_COMPACT_SIZE)
+    {
+        LogPrint("net", "Header error: size too large (%u bytes)\n", hdr.nMessageSize);
+        return -1;
+    }
```

Comment:
With the new approach, we achieved the goal of **detecting invalid messages earlier** (before passing them to the higher-level thread for processing), but we are also defining a new handling of errors.

Let's consider the following checks:
- **Valid message start**
- **Header validity**
- **Checksum**

**Before** this PR, they were performed by the `ProcessMessages()` function, and in case of failure, the **message was** just **discarded**. **Now**, this PR moves these checks to the `threadSocketHandler()` level, and in case of failure, the **peer** is **disconnected**.

As emerged in other discussions recently, that could be a problem because a failed check (e.g. invalid checksum) isn't necessarily caused by a malicious actor. It's very unlikely, but some network errors can silently "bypass" the TCP checksum and reach the application layer, **eventually causing the ban of honest peers**.

I like moving checks earlier in the application flow, but given the reasons explained above, I would rather keep the original behavior in terms of rejected messages, without banning nodes that may be honest.

Reply:
@JackPiri after reviewing further the PR I would propose the following behavior:

1. Inside `readHeader()`:
- In case of exception, return -1 and then disconnect (already done)
- In case of wrong message start, return -1 and then disconnect (to be adjusted)
- In case of size exceeded, return -1 and then disconnect (already done)
- No call to `hdr.HasValidCommand()` (see next sections)

2. Restore the previous behavior of `readData()`:
- Remove checksum computation

3. If message is complete():
- Validate the checksum (maybe the related code can be encapsulated into the `CNetMessage` class)?
- Call `hdr.HasValidCommand()`
- If any of the previous two checks fail, discard the message and count it in the statistics as "unknown" message (using `NetMsgType::OTHER`)
- If the two checks pass, then continue as usual.


I think that this should give us the same behavior as before and hopefully simplify a bit the readability of this section of code. What do you think?

Reply:
This should be addressed with commit afcb3a6cf7cf2200eedf25b020aae0f034a75e03.